ReactメモアプリSPA版の課題提出

yokomaru opened the pull request:
# 課題URL
https://bootcamp.fjord.jp/products/22596

# ReadMe
https://github.com/yokomaru/react-spa-memo-app/blob/db6c6beb780ccae6c03befe57cc05fea06cbf14c/README.md

yokomaru committed:
add: npm install

yokomaru committed:
add: Create basic CRUD functions using static data

一旦静的なデータを用いてアプリ全体を作成
- コンポーネントは以下で設計
　- MemoApp
        - MemoIndex
           - Memo
        - MemoEditor
           - MemoTextArea
- ステートは仮置き
     - memos
     - editingMemo
     - text

yokomaru committed:
fix: Move the delete and update buttons in the edit screen to the component above.

- propsの量が多いため、MemoTextAreaからMemoEditorに移動させる
- 不要な引数も削除
- console.logが残っていたため合わせて削除

yokomaru committed:
fix: Change editingMemoState to editingMemoID

textとeditingMemoでcontentの内容を被って持ってしまって２重管理になりそうなので、editingMemoIDのみ持たせるようにする

yokomaru committed:
fix: Do not use static data

MEMO配列を削除する(Strageからデータを持ってくる処理は別途作成)

yokomaru committed:
fix: Modify so that each file becomes one component

課題要件を満たすためにコンポーネントごとにファイルを分け、使用する各コンポーネントでimportするように修正

yokomaru committed:
fix: rename handleMemoClick to selectMemo

メモ要素をclickした時の動作のみに使用する関数名に見えるが、別途使用しているためメモの選択という関数名に修正

yokomaru committed:
fix: Separate list view and edit view

- 常に編集画面が表示になっていたが、要件的にはtextareaの表示切り替えが必要だったので追加
- 不要なstate変更を削除

yokomaru committed:
Fix: Replaced memo operations with useReducer

- useReducerを利用してmemoをCRUDするように修正
- 別ファイルに切り出した
- 引数の変更が必要な箇所を修正、合わせてpropsも修正

yokomaru committed:
add: Retrieving data from localStorage and setting values

- データの永続化のためにlocalStorageを使用
- 必要な処理をまずは追加

yokomaru committed:
add: Wrap the process that handles localStorage

- localStorage.jsにlocalStorageの処理を移し、Appから引数を渡すだけにする
- nextIdではなくmaxIdを取得しreducer内でインクリメントする

yokomaru committed:
fix: maxId is managed by useRef.

- コンポーネントに情報を記憶させたいので、useRefを使用

yokomaru committed:
add: prettier and fix style

- prettierの導入と実行
- 初期生成ファイルなどは対象外とする
- クォーテーションはシングルに統一

yokomaru committed:
fix: Adjusting CSS and HTML styles

- htmlの言語を日本語にする
- titleの修正
-  buttonで一覧表示している メモをaタグに置き換え
- + ボタンもaタグに置き換え
- < ul>< li> の ・も削除
- 背景色(body)を灰色にする
- textIndexとtextEditorを囲むdivを真ん中に寄せる
- textIndexとtextEditorを囲むdivを白にする
- textIndexとtextEditorを右と左に分ける

yokomaru committed:
fix: use a tag instead of a button to switch between memos.

- buttonタグを利用していたところをaタグに変更
- リンクを遷移するデフォルト動作を止めるためにe.preventDefault()を使う

yokomaru committed:
add: use useMemo

- メモの一覧をキャッシュするためにuseMemoを追加

yokomaru committed:
fix: Corrected incorrect naming of local storage operation processing.

- setがfromでgetがtoになっていたので修正

yokomaru committed:
fix: Delete memo title

- 要件がcontentのみで良さそうだったためtitleを削除

yokomaru committed:
fix: Readme

yokomaru committed:
fix: README.md

- lintの対応もれ

yokomaru committed:
fix: Unify memo update wording to “update”

- 更新はupdateを使用していたが、一部chengedに一部なっていたため修正

yokomaru commented on package.json:
```diff
@@ -22,6 +23,7 @@
     "eslint-plugin-react-hooks": "^5.2.0",
     "eslint-plugin-react-refresh": "^0.4.20",
     "globals": "^16.2.0",
+    "prettier": "^3.6.2",
```
ESLint と Prettier の両方でチェックしてから提出とあったので追加でインストールしました

yokomaru committed:
localStorage関連のバグを修正

- 名称と関数の中身が逆になっていた
- 上記によりuseEffect内で予期せずsetしており、useEffectの使い方に反した使い方をしていたためuseEffectからsetItemsを削除
- 各イベントハンドラにてsetするように修正
- changed更新漏れも対応

yokomaru committed:
fix: Readjust CSS styles

- 不要な要素を削除

yokomaru commented on src/utils/localStorage.js:
```diff
@@ -0,0 +1,7 @@
+export function setItemsToLocalStorage(key, array) {
+  localStorage.setItem(key, JSON.stringify(array));
+}
+
+export function getItemsFromLocalStorage(key) {
+  return JSON.parse(localStorage.getItem(key));
+}
```
try-catchを追加しようと思ったのですが、下記検討の末、どのようなエラーに対応するべきか現時点だと検討できず、一旦入れておりません。

- getItem
   - keyがない場合nullが返り、JSON.parseの結果もnullになるが、受け取り先で`getItemsFromLocalStorage('memo')  || []` としているため特に不要という判断をしました。

- setItem
    - 容量がいっぱいな場合(上限5MB)などにエラーになりうるかな？と思ったのですが、今回はテキストデータのみなので可能性は低いかな？と思っております。

yuuu replied:
そうですね。ここは極論「DevTool等でローカルストレージにJSONではない文字列を書かれる」可能性があるので、ちゃんと対策するなら「parseに失敗したらローカルストレージに `{}` を書いておく」といった対処方法が考えられます。

が、このプラクティスでそこまでのクオリティは求めていないのでマストではありません 👍

yokomaru replied:
コメントありがとうございます！
undefinedをparseするとエラーになることがわかったので、`SyntaxError`ならエラーメッセージ表示＆`[]`を返すように修正いたしました！

[82ac699](https://github.com/yokomaru/react-spa-memo-app/pull/1/commits/82ac699005733b61f338ac73ab0d956bd7a7f070)

yokomaru committed:
Update README.md

yokomaru committed:
fix: Readjust CSS styles

- 選択済みのaタグをクリックできないように修正

yokomaru committed:
fix: remove unnecessary line breaks README

- lintの修正対応

yuuu commented on src/App.jsx:
```diff
@@ -1,35 +1,112 @@
+  const [text, setText] = useState();
+
+  useEffect(() => {
+    maxId.current = Math.max(-1, ...memos.map((m) => m.id)) + 1;
```
`useEffect` で `memos` に更新が発生するたびに `maxId` を書き換えるのであれば、 `useRef` を使う必要はなく「 `useState` を使う」もしくは「 `handleAddButtonClick` で都度 `maxId` を計算する」という方法で良いように見えました 👀

yokomaru replied:
ご指摘ありがとうございます！

当初はmaxIDはレンダーとは関係ない値だと思っていたため、 useEffect + useRef を使って常に最大値を保存しておこうと思っていたのですが、メモが変わるたびに常に同期しなくてはならず、useRefの使い方としては適切ではなさそうでした💦

[Reactの流儀](https://ja.react.dev/learn/thinking-in-react#step-3-find-the-minimal-but-complete-representation-of-ui-state) にもあるように、maxId は memos から常に導出可能な値でしたので、今回の用途ではstateにはせずにhandleAddButtonClick の中で都度計算する形に修正いたします。

[867dbae](https://github.com/yokomaru/react-spa-memo-app/pull/1/commits/867dbae9b8f8fd65f9931da511168b9140712232) にて対応いたしました。

yuuu commented on src/App.jsx:
```diff
@@ -1,35 +1,112 @@
+
+  function handleMemoClick(e, memo) {
+    e.preventDefault();
+    setEditingMemoID(memo.id);
```
選択中のメモの `id` と `content` を別々のstateとしていますが、大半がセットで更新している状況なので `memo` オブジェクト自体をstateとして保持するようにしてはどうでしょう？

yokomaru replied:
MemoTextAreaへのonChangeに対応するstateとしてtextを使用していたのですが、オブジェクトの中身を書き換えるような形にすれば問題ないことがわかりました！(むしろ分けることで内容が乖離する可能性がある)
ありがとうございます。

以下で対応いたしました。
[aece7c6](https://github.com/yokomaru/react-spa-memo-app/pull/1/commits/aece7c630c37aec2e5c11fa27a286ae581f2ef49)
[4cc97cf](https://github.com/yokomaru/react-spa-memo-app/pull/1/commits/4cc97cf89dde6e913212526f6cc546113f0a2da4)

yuuu commented on src/App.jsx:
```diff
@@ -1,35 +1,112 @@
+  function handleUpdateButtonClick(editingMemoID, text) {
+    const trimmed = text.trim();
+    if (trimmed === '') {
+      alert('空白では入力できません');
```
バリデーションがしっかりしていると、ユーザビリティが上がって良いですね 👍

yuuu commented on src/App.jsx:
```diff
@@ -1,35 +1,112 @@
+    }
+    const updatedMemo = {
+      id: editingMemoID,
+      content: trimmed,
```
宣言の時点で `trimmed` という変数名を `content` 、 `editingID` を `id` にしておくとここの記述を簡略化できます。

```javascript
const updatedMemo = { id, content };
```

そんなシンプルな変数名で良いのか？という意見もありますが、他と区別する必要がないなら問題ないと思っています。
参考: https://bootcamp.fjord.jp/pages/avoid-too-specific-var-name

yokomaru replied:
ドキュメントの共有ありがとうございます。特にhandle関数内についてはmemoは1個しか登場しないため、シンプルな変数を使用するように修正いたしました。

[a927a59](https://github.com/yokomaru/react-spa-memo-app/pull/1/commits/a927a59b970d796335145f415566a4700068442c)
[29624cc](https://github.com/yokomaru/react-spa-memo-app/pull/1/commits/29624cc9448b3da5c8439b5c2c193f54079da475)

yuuu commented on src/App.jsx:
```diff
@@ -1,35 +1,112 @@
+    setEditingMemoID();
+  }
+
+  if (editingMemoID == null) {
```
ここは `==` ではなく `===` を使った方が良いですね。

yokomaru replied:
ご指摘ありがとうございます！

[57e606d](https://github.com/yokomaru/react-spa-memo-app/pull/1/commits/57e606d9fc5a5c505dbe6ca7c6acbe3f112c714c)にて厳密等価演算子に変更いたしました。

yuuu commented on src/App.jsx:
```diff
@@ -1,35 +1,112 @@
+      <>
+        <div className="main">
+          <div className="memo-index full">
+            <MemoIndex
```
このコンポーネント呼び出しが2回登場するので、

- divタグのクラス名 `full` `half` を `editingMemoID` の有無に応じて切り替える
- `editingMemoID` が存在するときのみ `MemoEditor` を表示する

としたほうが重複量は少なくて済みそうです。
`editingMemoID === null` の判定は2回になってしまいますが、そちらの方がコードとしてはスッキリしそう。

yokomaru replied:
おっしゃる通り、判定が2回になるものの全体のコード量はかなり減りました！

[eb65f46](https://github.com/yokomaru/react-spa-memo-app/pull/1/commits/eb65f46da16fc8d20bc8545fb57b83a47a0bc955) で修正いたしました！

yuuu commented on src/reducers/memosReducer.js:
```diff
@@ -0,0 +1,22 @@
+export default function memosReducer(memos, action) {
```
reducerを使ってステート管理すると、バケツリレーを減らせたり、ステートに関するロジックを切り出しやすくなったりといった効果があって良いですね 👍

yokomaru committed:
fix: Change the equality operator (==) to the strict equality operator (===)

意図しない挙動を防ぐため、nullとの比較は厳密等価演算子（===）にする

yokomaru committed:
fix: Move the maxId retrieval process to handleAddButtonClick

- useEffect内でuseRefを用いていたが、memosの更新を元に(レンダリングのたびに)データを更新していたためuseRefを用いる意味がないので、handle内で都度最大を更新する方法に変更(useEffectも削除)
- handleAddButtonClickにsetItemsToLocalStorageを移動
- handleDeletedButtonClickにもsetItemsToLocalStorageを追加

yokomaru committed:
fix: Put setItemsToLocalStorage in useEffect

- handleClick内で更新されたmemosは更新前のmemosのため、setItemsToLocalStorageをhandleClick内に置くとデータがおかしくなるのでuseEffectでmemosの状態が変わったらlocalStorageを更新しに行くように修正した

- update時、delete時にsetTextとsetEdittingMemoIDを更新していなかったので更新した

yokomaru committed:
fix: Change the value to be compared from null to undefined.

- edittingMemoIDが空の場合nullではなくundefinedのため比較する値を修正

yokomaru committed:
fix: Change to simple variable names (id, content)

- 以下を参考に他と区別が不要な場合はシンプルな変数名をつける
    - https://bootcamp.fjord.jp/pages/avoid-too-specific-var-name
- memoEditorに渡すidもシンプルにする(編集中のidしか渡さないため)

yokomaru committed:
fix: editingMemoIDでの分岐

- memo-indexの halfとfullをeditingMemoIDの有無の分岐で表示する
- editingMemoID が存在するときのみ MemoEditor を表示する

yokomaru committed:
fix: 編集中のメモをオブジェクトで持つようにする

- setTextとsetEdittingMemoIDで分けていたが、統合しオブジェクトで管理するように修正

yokomaru committed:
fix: Enclose JSON.parse in try-catch

- try-catchを追加しJSON.parseに失敗したらfalseを返して[]が代入されるように修正

yokomaru committed:
fix: Change to spread syntax

- edittingMemoにsetするオブジェクトをスプレッド構文を用いてcontentだけ上書きする

yokomaru committed:
Merge branch 'my-react-spa-memo-app' of https://github.com/yokomaru/react-spa-memo-app into my-react-spa-memo-app

yokomaru committed:
fix: Simplify variables within the handle function

- 他と区別する必要がないのでedittingMemoをmemoに変更

yokomaru commented on src/App.jsx:
```diff
@@ -1,35 +1,90 @@
+  const [editingMemo, setEditingMemo] = useState();
+
+  useEffect(() => {
+    setItemsToLocalStorage('memos', memos);
```
各hundle内で`setItemsToLocalStorage`をしていたのですが、更新前のmemosで更新してしまっていたためストレージとmemosの同期がうまくいっていないことがわかりました💦

`useEffect`を用いて`memos`に変化があった後に`setItemsToLocalStorage`を変更するようにしてみたのですが、`useEffect`の使い道としてはあっておりますでしょうか？

また、`useEffect`を使う以外での解決方法もあるのでしょうか？(更新用関数(`(n) => n + 1`) 的なものも使えるのだろうか？と思ったのですが、`useReducer`を使っているのもありいまいち適用方法がわかりませんでした😵)

yuuu replied:
いえ、この方法がベターだと思いました 👍 `memos` が更新されたタイミングでローカルストレージに書き出したいので。

`App` コンポーネントにロジックが多くなっていることが気になるのであれば、 `useReducer` `useEffect` `handleAddButtonClick` `handleUpdateButtonClick` `handleDeleteButtonClick` の定義をまとめた `useMemos` のような名前のカスタムフックを作るのも手ですね。

yokomaru replied:
`useEffect`での方法がベターとのこと、理解いたしました！(ローカルストレージへの更新は`memos`の状態の変化によって起こる副作用なので`useEffect`の使用法として問題ないという風に理解いたしました)

また、カスタムフックにつきましても実装いたしました！
[ef37b4a](https://github.com/yokomaru/react-spa-memo-app/pull/1/commits/ef37b4a1772571af89f56612df3171d53cb6be99)

ご提案いただいた`useReducer`(`memoReducer.js`自体はそのまま別ファイル)、`useEffect`、`handleAddButtonClick`、`handleUpdateButtonClick `、`handleDeleteButtonClick`に加え、`editingMemo`ステートも`useMemos`に移動しました。

`editingMemo`ステートを含めた理由としては以下になります。

- どちらもメモに関するステートなので、同じカスタムフックに切り出しても違和感がない(`useMemos`という命名でも問題ないはなさそう)と感じた
    - ただ、`memos`と`edittingMemo`の目的が違うもの(メモというデータの保存と現在編集中のメモを保持する)なので、同じカスタムフックに存在していいのかいまいち自信が持てない状態
- `handle`メソッド内にて`setEdittingMemo`を更新しているため切り出すのが難しかった
   - 以下のように`App`にて`editingMemo`を定義し`useMemos`に渡すことも考えたが、`App`で定義しているステートを`useMemos`で更新することに違和感を感じた
```jsx
  // App.jsx
  const [editingMemo, setEditingMemo] = useState();
  const {
    memos,
    handleMemoClick,
    handleAddButtonClick,
    handleUpdateButtonClick,
    handleDeleteButtonClick,
   } = useMemos(setEditingMemo);
```

yuuuさん的にこのカスタムフックは問題がありそうでしょうか？
ご意見をいただけますと幸いです🙇‍♀️

yuuu replied:
修正検討ありがとうございます 🙏 なるほど。確かに、 `handle` メソッドを介して2つの情報を操作しているので同じカスタムフックにせざるを得ないというのは理解できます。

しかし、別々のステートのものを一つのカスタムフックにするとそのカスタムフックを別のアプリ等に流用しづらくなるため避けたいですね。一つの手として、子コンポーネントに渡すコールバック関数は別途作成して、その中でカスタムフックが返す `hendle` メソッドと `setEdittingMemo` を呼ぶようにすると、カスタムフックが持つステートを1つにできるはずです。

全体としては、冗長になってしまいますが、その方がカスタムフックの使い回しがしやすくなるので望ましいと思いました 👍

yokomaru replied:
@yuuu
返信ありがとうございます！

> 別々のステートのものを一つのカスタムフックにするとそのカスタムフックを別のアプリ等に流用しづらくなるため避けたい

なるほど、、！せっかくカスタムフックにしたのに汎用性がなくなるという問題が出てくるのですね！ここら辺の判断が難しかったのでとても勉強になりました。

> 子コンポーネントに渡すコールバック関数は別途作成して、その中でカスタムフックが返す hendle メソッドと setEdittingMemo を呼ぶようにする

こちら、以下のようなイメージで役割を切り分ける認識なのですが、合っておりますでしょうか？
念の為確認させてください！

- App.jsx(ブラウザからの各イベントに反応する処理。memoステートとedittingMemoステートの更新がまとまっている。このメソッドを子コンポーネントに渡す)
    - handleMemoClick
    - handleAddButtonClick
    - handleUpdateButtonClick
    - handleDeleteButtonClick

- useMemo.jsx(Memoステートのみに特化したカスタムフック。以下メソッドはdispatchをラップしたメソッド)
    - addMemo
    - updateMemo
    - DeleteMemo

お手隙でご確認お願いいたします🙇‍♀️

yuuu replied:
はい、上記で認識合っています 👍

yokomaru replied:
コメントありがとうございます！
以下で実装いたしました！

[04f7b1c](https://github.com/yokomaru/react-spa-memo-app/pull/1/commits/04f7b1cf1ec6f5e66f60bbbe8ad5a182ab49245e)
[9d4b0b3](https://github.com/yokomaru/react-spa-memo-app/pull/1/commits/9d4b0b394bb4f3edec2256ed6904528d115d2a3a)

yuuu commented on src/App.jsx:
```diff
@@ -1,35 +1,90 @@
+    }
+    const updatedMemo = {
+      id: memo.id,
+      content: content,
```
ここは次のように省略できますね

```suggestion
      content,
```

yokomaru replied:
キーと変数名が同じ場合省略記法で記載できるのですね！とても勉強になりました。

[f14bcb8](https://github.com/yokomaru/react-spa-memo-app/pull/1/commits/f14bcb8a3de57769b0b23d7f7c371cd6515012ca) にて修正いたしました！

yokomaru committed:
fix: Application of abbreviated notation

- handleUpdateButtonClick内で更新用のmemoのオブジェクトにて、キーと変数が同じ名前のため省略記法を適用する

yokomaru committed:
add: Add custom hook "useMemos "

- メモに関する状態を処理を操作する処理をカスタムフックに切り出した
    - 一緒に更新されることが多いため、memosとeditingMemo両方移動させた

yokomaru committed:
fix: Correct LocalStorage to localStorage.

- ファイルがキャメルケースで定義されていたがimport時に大文字になっていたため修正

yokomaru committed:
fix: Modify UseMemo to a custom hook that handles only Memo

- 別々のステートを一つのカスタムフックにしていたことでカスタムフックが流用しにくくなっていたためEditting MemoはAppで管理し、UseMemoをMemoのみ扱うカスタムフックに修正

yokomaru committed:
fix: Fixed how memo objects are passed

- 省略を適用するために、addMemoとupdateMemoのmemoの渡し方を修正
